bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java

```diff
@@ -267,18 +242,40 @@ public void run() {
       }
     }, executorService);
 
+    int timeout = retryOptions.getReadPartialRowTimeoutMillis();
+
+    int streamingBufferSize = retryOptions.getStreamingBufferSize();
+    ResponseQueueReader responseQueueReader =
+        new ResponseQueueReader(timeout, streamingBufferSize, streamingBufferSize, readRowsCall);
+
     StreamingBigtableResultScanner resultScanner =
-        new StreamingBigtableResultScanner(
-          reservedChannel,
-          retryOptions.getStreamingBufferSize(),
-          retryOptions.getReadPartialRowTimeoutMillis(),
-          cancellationToken);
+        new StreamingBigtableResultScanner(reservedChannel, responseQueueReader, cancellationToken);
 
     clientCallService.asyncServerStreamingCall(
         readRowsCall,
         request,
-        new ReadRowsStreamObserver(resultScanner));
+        streamingBufferSize,
+        createClientCallListener(resultScanner));
 
     return resultScanner;
   }
+
+  private ClientCall.Listener<ReadRowsResponse> createClientCallListener(
+      final StreamingBigtableResultScanner resultScanner) {
+    return new ClientCall.Listener<ReadRowsResponse>() {
+      @Override
+      public void onMessage(ReadRowsResponse readRowResponse) {
+        resultScanner.addResult(readRowResponse);
```

Comment:
instead of what u have, can we do this:

  private ClientCall.Listener<ReadRowsResponse> createClientCallListener(
      final StreamingBigtableResultScanner resultScanner, final ClientCall<?, ReadRowsResponse> call) {
    return new ClientCall.Listener<ReadRowsResponse>() {
      @Override
      public void onMessage(ReadRowsResponse readRowResponse) {
        resultScanner.addResult(readRowResponse);
        // if (not end of result) {
        //   call.request(x) where x = capacity - available;
        // }
        call.request(x);
      }